File: candas-backend/src/main/java/com/candas/candas_backend/service/DistribuidorService.java
package com.candas.candas_backend.service;

import com.candas.candas_backend.dto.DistribuidorDTO;
import com.candas.candas_backend.entity.Distribuidor;
import com.candas.candas_backend.exception.ResourceNotFoundException;
import com.candas.candas_backend.repository.DistribuidorRepository;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@Transactional
public class DistribuidorService {

    private final DistribuidorRepository distribuidorRepository;

    public DistribuidorService(DistribuidorRepository distribuidorRepository) {
        this.distribuidorRepository = distribuidorRepository;
    }

    public Page<DistribuidorDTO> findAll(Pageable pageable) {
        return distribuidorRepository.findAll(pageable).map(this::toDTO);
    }

    public DistribuidorDTO findById(Long id) {
        Distribuidor distribuidor = distribuidorRepository.findById(id)
            .orElseThrow(() -> new ResourceNotFoundException("Distribuidor", id));
        return toDTO(distribuidor);
    }

    public List<DistribuidorDTO> search(String query) {
        if (query == null || query.trim().isEmpty()) {
            return List.of();
        }
        List<Long> ids = distribuidorRepository.searchIds(query.trim());
        if (ids.isEmpty()) {
            return List.of();
        }
        return distribuidorRepository.findAllById(ids).stream()
            .map(this::toDTO)
            .collect(java.util.stream.Collectors.toList());
    }

    public DistribuidorDTO create(DistribuidorDTO dto) {
        Distribuidor distribuidor = toEntity(dto);
        distribuidor.setActiva(true);
        return toDTO(distribuidorRepository.save(distribuidor));
    }

    public DistribuidorDTO update(Long id, DistribuidorDTO dto) {
        Distribuidor distribuidor = distribuidorRepository.findById(id)
            .orElseThrow(() -> new ResourceNotFoundException("Distribuidor", id));
        
        distribuidor.setNombre(dto.getNombre());
        distribuidor.setCodigo(dto.getCodigo());
        distribuidor.setEmail(dto.getEmail());
        if (dto.getActiva() != null) {
            distribuidor.setActiva(dto.getActiva());
        }
        
        return toDTO(distribuidorRepository.save(distribuidor));
    }

    public void delete(Long id) {
        Distribuidor distribuidor = distribuidorRepository.findById(id)
            .orElseThrow(() -> new ResourceNotFoundException("Distribuidor", id));
        distribuidor.setActiva(false);
        distribuidorRepository.save(distribuidor);
    }

    public DistribuidorDTO buscarOCrear(String nombre, String codigo) {
        // Buscar por código primero si está disponible
        if (codigo != null && !codigo.trim().isEmpty()) {
            return distribuidorRepository.findByCodigoIgnoreCase(codigo.trim())
                .map(this::toDTO)
                .orElseGet(() -> {
                    // Crear nueva si no existe
                    DistribuidorDTO nuevoDTO = new DistribuidorDTO();
                    nuevoDTO.setNombre(nombre != null ? nombre.trim() : "Sin nombre");
                    nuevoDTO.setCodigo(codigo.trim());
                    nuevoDTO.setActiva(true);
                    return create(nuevoDTO);
                });
        }
        
        // Buscar por nombre si no hay código
        if (nombre != null && !nombre.trim().isEmpty()) {
            return distribuidorRepository.findByNombreIgnoreCase(nombre.trim())
                .map(this::toDTO)
                .orElseGet(() -> {
                    // Crear nueva si no existe
                    DistribuidorDTO nuevoDTO = new DistribuidorDTO();
                    nuevoDTO.setNombre(nombre.trim());
                    nuevoDTO.setActiva(true);
                    return create(nuevoDTO);
                });
        }
        
        // Si no hay nombre ni código, crear con nombre por defecto
        DistribuidorDTO nuevoDTO = new DistribuidorDTO();
        nuevoDTO.setNombre("Sin nombre");
        nuevoDTO.setActiva(true);
        return create(nuevoDTO);
    }

    private DistribuidorDTO toDTO(Distribuidor distribuidor) {
        DistribuidorDTO dto = new DistribuidorDTO();
        dto.setIdDistribuidor(distribuidor.getIdDistribuidor());
        dto.setNombre(distribuidor.getNombre());
        dto.setCodigo(distribuidor.getCodigo());
        dto.setEmail(distribuidor.getEmail());
        dto.setActiva(distribuidor.getActiva());
        return dto;
    }

    private Distribuidor toEntity(DistribuidorDTO dto) {
        Distribuidor distribuidor = new Distribuidor();
        distribuidor.setNombre(dto.getNombre());
        distribuidor.setCodigo(dto.getCodigo());
        distribuidor.setEmail(dto.getEmail());
        return distribuidor;
    }
}

File: candas-frontend/src/lib/api/distribuidor.service.ts
import type { Distribuidor, DistribuidorPage } from '@/types/distribuidor'
import { openapiClient, handleResponse } from './openapi-client'

export interface DistribuidorFindAllParams {
  page?: number
  size?: number
  search?: string
  nombre?: string
  codigo?: string
  activa?: boolean
}

export const distribuidorService = {
  async findAll(params: DistribuidorFindAllParams = {}): Promise<DistribuidorPage> {
    const { page = 0, size = 20, search, nombre, codigo, activa } = params
    return handleResponse(
      openapiClient.GET('/api/v1/distribuidores', {
        params: {
          query: {
            pageable: { page, size },
            search,
            nombre,
            codigo,
            activa,
          },
        },
      })
    ) as any
  },

  async findById(id: number): Promise<Distribuidor> {
    return handleResponse(
      openapiClient.GET('/api/v1/distribuidores/{id}', {
        params: {
          path: { id },
        },
      })
    ) as any
  },

  async create(dto: Distribuidor): Promise<Distribuidor> {
    return handleResponse(
      openapiClient.POST('/api/v1/distribuidores', {
        body: dto as any,
      })
    ) as any
  },

  async update(id: number, dto: Distribuidor): Promise<Distribuidor> {
    return handleResponse(
      openapiClient.PUT('/api/v1/distribuidores/{id}', {
        params: {
          path: { id },
        },
        body: dto as any,
      })
    ) as any
  },

  async delete(id: number): Promise<void> {
    return handleResponse(
      openapiClient.DELETE('/api/v1/distribuidores/{id}', {
        params: {
          path: { id },
        },
      })
    ) as any
  },

  async search(query: string): Promise<Distribuidor[]> {
    return handleResponse(
      openapiClient.GET('/api/v1/distribuidores/search', {
        params: {
          query: { query },
        },
      })
    ) as any
  },

  async buscarOCrear(nombre?: string, codigo?: string): Promise<Distribuidor> {
    return handleResponse(
      (openapiClient as any).POST('/api/v1/distribuidores/buscar-o-crear', {
        body: { nombre, codigo } as any,
      })
    ) as any
  },
}
